src/lib/ldap/base.h

```diff
@@ -483,6 +487,57 @@ typedef struct {
 	fr_ldap_result_code_t	ret;		//!< Return code of bind operation.
 } fr_ldap_bind_auth_ctx_t;
 
+typedef struct ldap_filter_s ldap_filter_t;
+
+/** Types of parsed LDAP filter nodes
+ */
+typedef enum {
+	LDAP_FILTER_NODE		= 0,		//!< The filter node is an individual one
+							//!< to be evaluated against an attribute.
+	LDAP_FILTER_GROUP				//!< The filter node is a parent of a group
+							//!< which will be combined using a logical operator.
+} ldap_filter_type_t;
+
+/** Logical operators for use in LDAP filters
+ */
+typedef enum {
+	LDAP_FILTER_LOGIC_AND		= 1,
+	LDAP_FILTER_LOGIC_OR,
+	LDAP_FILTER_LOGIC_NOT
+} ldap_filter_logic_t;
+
+/** Operators for use in LDAP filters
+ */
+typedef enum {
+	LDAP_FILTER_OP_UNSET		= 0,		//!< Attribute not set yet
+	LDAP_FILTER_OP_EQ,				//!< Attribute equals value
+	LDAP_FILTER_OP_SUBSTR,				//!< Attribute matches string with wildcards
+	LDAP_FILTER_OP_PRESENT,				//!< Attribute present
+	LDAP_FILTER_OP_GE,				//!< Attribute greater than or equal to value
+	LDAP_FILTER_OP_LE,				//!< Attribute less than or equal to value
+	LDAP_FILTER_OP_BIT_AND,				//!< Bitwise AND comparison
+	LDAP_FILTER_OP_BIT_OR				//!< Bitwise OR comparison
+} ldap_filter_op_t;
+
+/** Structure to hold parsed details of LDAP filters
+ */
+struct ldap_filter_s {
+	ldap_filter_type_t	filter_type;		//!< Type of this filter node
+	union {
+		struct {
+			ldap_filter_logic_t	logic_op;	//!< Logical operator for this group
+			ldap_filter_t		*child;		//!< First child of this group
+		};
+		struct {
+			char			*attr;		//!< Attribute for the filter node
+			ldap_filter_op_t	op;		//!< Operator to be used for comparison
+			fr_value_box_t		*value;		//!< Value to compare with
+		};
+	};
+	ldap_filter_t		*next;			//!< Next node in a group
```

Comment:
I know this is a pain, but we're trying to move everything to dlists.  Could you setup a type wrapper for dlists and use those instead?